app/ModelFunctions/LDAPFunctions.php

```diff
@@ -0,0 +1,377 @@
+<?php
+
+namespace App\ModelFunctions;
+
+use App\Exceptions\LDAPException;
+use App\Models\Configs;
+use App\Models\Logs;
+
+class LDAPFunctions
+{
+	/* @var resource $con holds the LDAP connection */
+	protected $con = null;
+
+	/* @var int $bound What type of connection does already exist? */
+	protected $bound = 0; // 0: anonymous, 1: user, 2: superuser
+
+	/* @var array user info queried from LDAP */
+	protected $user_info = [];
+
+	/**
+	 * Log debug message.
+	 *
+	 * @param string $method
+	 * @param string $line
+	 * @param string $text
+	 */
+	protected function _debug(string $method, string $line, string $text)
+	{
+		if ($this->getConf('debug', '0')) {
+			Logs::notice($method, $line, $text);
+		}
+	}
+
+	/**
+	 * Wraps around ldap_search, ldap_list or ldap_read depending on $scope.
+	 *
+	 * @param resource   $link_identifier
+	 * @param string     $base_dn
+	 * @param string     $filter
+	 * @param string     $scope           can be 'base', 'one' or 'sub'
+	 * @param array|null $attributes
+	 * @param int        $attrsonly
+	 * @param int        $sizelimit
+	 *
+	 * @return resource
+	 */
+	protected function _ldapsearch($link_identifier, $base_dn, $filter, $scope = 'sub', $attributes = null,
```

Comment:
I decided to use LDAPException only once if the connection to the LDAP server cannot be established.

The rest of the code would not fire Exceptions if the PHP LDAP functions do not fire them.
If the only consequence of omitting the @ operator now that you ask me to document  all the Exceptions which might be fired by the library functions in your code then I consider it a wast of time to remove the @ operator.

Reply:
Another solution would be to wrap the whole function definition into a `try{ ... } catch(\ExceptionThrownByLibarary $e) {...}` block and then only throw your custom `LDAPException` _and_ attach the original exception as the predecessor exception.

Reply:
This is all no solution, since the Exceptions fired by ldap_bind cannot be caught. So the only solution for not breaking the application is the use of the @-operator.

So the request to eliminate the @-operator in my code is simply nonsence. The @-operator is needed if you want a stable application. That is simply a fact. I figured this out long ago and I spend a lot of time verifying this fact with php 8.1.5 again only because you were not knowing what you are asking for.

So it was a waste of time trying to remove the @-operator. It is simply required until someone changes the PHP ldap-functions.

I would appreciate it if you can avoid such nonsense.requests in future.

Reply:
> This is all no solution, since the Exceptions fired by ldap_bind cannot be caught. So the only solution for not breaking the application is the use of the @-operator.

Why should it be impossible to catch exceptions thrown by `ldap_bind` and related functions? As far as I knoe the PHP language does not support a feature like "uncaughtable" exceptions. If you really have encountered such an exception, this would imply a bug in the PHP engine and I highly doubt that.

A MWE which illustrates this problem would be highly welcome.

> So the request to eliminate the @-operator in my code is simply nonsence. The @-operator is needed if you want a stable application. That is simply a fact. I figured this out long ago and I spend a lot of time verifying this fact with php 8.1.5 again only because you were not knowing what you are asking for.

The links to the RFC for PHP and the release notes about the `@`-operator are included in the first comment. The `@`-operator is legacy and it is just a question of time when it will be removed entirely from PHP. We have removed the `@`-operator from Lychee and it works perfectly. Unless you can illustrate your problem, I still need to ask you to change it.

Reply:
Look, if you think there are no "uncaughtable" exceptions in php then that is fine with me. Everybody can have its believes. For us who want stable code, do the testing and put in the @-operator.

I do not care if it a "bug" in PHP or a bug in the PHP ldap library. I put in the @-operator and try to solve other problems.

It is actually not a bug. It is a feature. If you cannot get your ldap_bind, the script does not run. It is very secure. It is just not the right way for modern web applications.

Its very simple to demonstrate. Just remove all the @-operators from my code and run the unit test I have written.
Now you will see ErrorExceptions. But how can that happen? There is a catch statement around all the library calls.

The next thing you should do is using mother google and there you will find comments from years ago reporting this feature. And the solution is always: use the @-operator.

Just because someone writes "the @-operator is legacy" you should not believe it. It is legacy if it is not needed any more. ldap_bind needs a @-operator if you do not want you application to crash if someone types in a wrong password.

There is no point in proving facts.

Reply:
> Its very simple to demonstrate. Just remove all the @-operators from my code and run the unit test I have written. Now you will see ErrorExceptions. But how can that happen? There is a catch statement around all the library calls.

If you refer to this kind of catch statement

https://github.com/hfr/Lychee/blob/6d8fff95bc4a88be7bd422c6e971e77f21e8d55a/app/ModelFunctions/LDAPFunctions.php#L78

it simply does not work, because you try to catch the wrong type of exception. `\ErrorException` does not inherit `\Exception`, because `\Exception` is not the parent class of all exceptions in PHP. This is `\Throwable`.

The top-level inheritance chain in PHP looks like this

                                        +--- \RuntimeException
                   +---  \Exception  ---+
                   |                    +--- \LogicException
    \Throwable  ---+
                   +---  \ErrorException

So in order to catch `\ErrorException` and `\Exception` within the same catch-`statement` you must use `\Throwable`

> Look, if you think there are no "uncaughtable" exceptions in php then that is fine with me. Everybody can have its believes.

"My believes" are substantiated by original, first-hand documentation which I posted you at the very beginning. 

> The next thing you should do is using mother google and there you will find comments from years ago reporting this feature. And the solution is always: use the @-operator.

The problem with mother Google and (I assume) StackOverflow is that they are usually a very bad source compared to original literature. They are just not authoritative. You will find a "proof" for everything on Google. Especially, in the field of programming language, if something has been posted years ago, i.e. in the "dark ages" of PHP, it probably won't be correct today.
 
> Just because someone writes "the @-operator is legacy" you should not believe it.

Except that this someone is a maintainer of the PHP core development team. I think this is amazing: You are willing to accept answers from "mother Google" without questioning, but a first-hand reference from the php.net site written by a PHP maintainer is "someone one should not believe".

> There is no point in proving facts.

Of course it is, because a fact isn't a fact until you have proven it or provided evidence for it. The reason why I asked you for a MWE was that I expected you would use `catch (\Exception $e)`.

So once again and for the last time: change your code.

Reply:
> > Its very simple to demonstrate. Just remove all the @-operators from my code and run the unit test I have written. Now you will see ErrorExceptions. But how can that happen? There is a catch statement around all the library calls.
> 
> If you refer to this kind of `catch`-statement
> 
> https://github.com/hfr/Lychee/blob/6d8fff95bc4a88be7bd422c6e971e77f21e8d55a/app/ModelFunctions/LDAPFunctions.php#L78
> 
> it simply does not work, because you try to catch the wrong type of exception. `\ErrorException` does not inherit `\Exception`, because `\Exception` is not the parent class of all exceptions in PHP. This is `\Throwable`.
> 
> The top-level inheritance chain in PHP looks like this
> 
> ```
>                                     +--- \RuntimeException
>                +---  \Exception  ---+
>                |                    +--- \LogicException
> \Throwable  ---+
>                +---  \ErrorException
> ```
> 
> So in order to catch `\ErrorException` and `\Exception` within the same `catch`-statement you must use `\Throwable`
> 
> > Look, if you think there are no "uncaughtable" exceptions in php then that is fine with me. Everybody can have its believes.
> 
> "My believes" are substantiated by original, first-hand documentation which I posted you at the very beginning.
> 
> > The next thing you should do is using mother google and there you will find comments from years ago reporting this feature. And the solution is always: use the @-operator.
> 
> The problem with mother Google and (I assume) StackOverflow is that they are usually a very bad source compared to original literature. They are just not authoritative. You will find a "proof" for everything on Google. Especially, in the field of programming language, if something has been posted years ago, i.e. in the "dark ages" of PHP, it probably won't be correct today.
> 
> > Just because someone writes "the @-operator is legacy" you should not believe it.
> 
> Except that this someone is a maintainer of the PHP core development team. I think this is amazing: You are willing to accept answers from "mother Google" without questioning, but a first-hand reference from the php.net site written by a PHP maintainer is "someone one should not believe".
> 
> > There is no point in proving facts.
> 
> Of course it is, because a fact isn't a fact until you have proven it or provided evidence for it. The reason why I asked you for a MWE was that I expected you would use `catch (\Exception $e)`.
> 
> So once again and for the last time: change your code.

Of course I am not referring to that catch-statement.

You are obviously very stubben.

There is no point in trying to prove facts to stubben people. It simply does not work.

You have the strange idea that the ErrorExeptions created by ldap_bind can be caught. The fact is that they can't. Do not ask me why not. I simply do not care.

The information about the class inheritance was simply wrong is you can read here: https://www.php.net/manual/en/class.exception.php

ErrorException is an Exception from PHP 7.2 on.

The only thing which works is to set up the good old error handler. But for me this is even more ugly as the @-operator.

But I have implemented it to prove my assumption that it might work. It does, its a fact.

So my suggestion is that we stay with the @-operator as long as it is around and see how error handling might have evolved when it is not supported any more.

By the way, I am not willing to accept any answer without testing them. For me only facts count and it is a fact, if it runs on my machine...

Reply:
> > If you refer to this kind of catch-statement [...]
>
> Of course I am not referring to that catch-statement. You are obviously very stubben. There is no point in trying to prove facts to stubben people. It simply does not work.

Until know you haven't specifically referred to any catch clause. Hence I wrote "if you refer ...". Also I did a typo, I didn't meant `ErrorException`, but simply `Error` this was my fault. But the rest is correct. See here https://www.php.net/manual/de/class.throwable.php and here https://www.php.net/manual/de/class.error.php.

Anyway, this is a last fair chance:

You are constantly insulting people and lecturing them. This is nothing, I would like to see any further. Either you watch your tone and do as you have been asked or this PR is not going to be merged. Any further misstep and this PR is going to be closed. As an ultima ratio we can also block you from this site.

Reply:
You simply do not understand. I do not care if you accept my PR or not.

If have my installations up and running and if you do not want this feature then that is fine with me. Since it is in GIT I can merge it in any time, lychee gets updated.

Trying to get the @-operator out of the code was completely waste of time, just because you did not do the tests in practice.

This is something I cannot change. I can write reliable code and I can find problems, but manage you attitude to facts is simply out of my scope.

Since the problem is completely on your side, I do not have any understanding for the following sentence: 

"Either you watch your tone and do as you have been asked or this PR is not going to be merged."

Isn't it about getting useful and reliable features implemented?

I cannot change that you got it wrong. Since it now gets all obvious you get nasty. Why?

It would be simply better, if you try to avoid nonsense requirements.

By the way if you really a maintainer of the PHP core development team as you have suggested then you should take this as a chance for an improvement. The good old error handler can handle the ldap_bind Exceptions, but you cannot catch them. There must be a reason for this somewhere. If you find it, you can improve the world at least a bit...